File: 110_Balanced_Binary_Tree.py
"""
Given a binary tree, determine if it is height-balanced.

For this problem, a height-balanced binary tree is defined
as a binary tree in which the depth of the two subtrees of
every node never differ by more than 1.
"""

#
# Key Points: Recursion.
#             If a tree is a balanced tree, its left and right subtrees
#             must both be balanced tree and their depth can't differ
#             more than 1
#
#             Therefore, define a new function isSubtreeBalanced(root).
#             If root is balanced, return its depth, else return -1
#
# Run time: 80 ms
#

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def isBalanced(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        def isSubtreeBalanced(root):
            if not root:
                return 0
            left = isSubtreeBalanced(root.left)
            right = isSubtreeBalanced(root.right)
            if left >=0 and right >= 0 and abs(left-right) <=1:
                return max(left, right) + 1
            else:
                return -1
        if isSubtreeBalanced(root) >= 0:
            return True
        return False

File: java/110_Balanced_Binary_Tree.java
// Given a binary tree, determine if it is height-balanced.

// For this problem, a height-balanced binary tree is defined as a binary tree in which the depth of the two subtrees of every node never differ by more than 1.


/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public boolean isBalanced(TreeNode root) {
        return getNodeInfo(root)[0] == 1;
    }
    public int[] getNodeInfo(TreeNode node) {
        int[] retValue = {1, -1};
        if (node == null){
            return retValue;
        }
        
        int[] leftInfo = getNodeInfo(node.left);
        int[] rightInfo = getNodeInfo(node.right);

        if ((leftInfo[0] == 1)&& (rightInfo[0]==1)&&( Math.abs(leftInfo[1] - rightInfo[1]) <= 1)){
            retValue[1] = Math.max(rightInfo[1], leftInfo[1]) + 1;
        }else{
            retValue[0] = 0;
        }
        return retValue;
    }
}